src/app/helpers/attachment/attachmentLoader.ts

```diff
@@ -0,0 +1,134 @@
+import {
+    binaryStringToArray,
+    concatArrays,
+    decodeBase64,
+    decryptMessage,
+    decryptSessionKey,
+    getMessage
+} from 'pmcrypto';
+
+import { getAttachment } from '../../api/attachments';
+import { MessageExtended } from '../../models/message';
+import { Attachment, AttachmentData } from '../../models/attachment';
+import { AttachmentsCache } from '../../hooks/useAttachments';
+import { Api } from '../../models/utils';
+
+// Reference: Angular/src/app/attachments/services/AttachmentLoader.js
+
+interface SessionKey {
+    data?: Uint8Array;
+    algorithm?: string;
+}
+
+// TODO: Handle isOutside()
+
+export const decrypt = async (
+    encryptedBinaryBuffer: ArrayBuffer,
+    sessionKey: SessionKey = {}
+): Promise<AttachmentData> => {
+    const encryptedBinary = new Uint8Array(encryptedBinaryBuffer);
+
+    try {
+        return await decryptMessage({
+            message: await getMessage(encryptedBinary),
+            sessionKeys: [sessionKey],
+            format: 'binary'
+        });
+    } catch (err) {
+        console.error(err);
+        throw err;
+    }
+};
+
+export const getRequest = ({ ID }: Attachment = {}, api: Api): Promise<ArrayBuffer> => {
+    // if (isOutside()) {
+    //     const decryptedToken = eoStore.getToken();
+    //     const token = $stateParams.tag;
+    //     return Eo.attachment(decryptedToken, token, ID);
+    // }
+
+    return api(getAttachment(ID));
+};
+
+export const getSessionKey = async (attachment: Attachment, message: MessageExtended): Promise<SessionKey> => {
+    // if (attachment.sessionKey) {
+    //     return attachment;
+    // }
+
+    const keyPackets = binaryStringToArray(decodeBase64(attachment.KeyPackets));
+    const options: any = { message: await getMessage(keyPackets) };
+
+    // if (isOutside()) {
+    //     options.passwords = [eoStore.getPassword()];
+    // } else {
+    // options.privateKeys = keysModel.getPrivateKeys(message.AddressID);
+    options.privateKeys = message.privateKeys;
+    // }
+
+    const sessionKey = await decryptSessionKey(options);
+
+    return sessionKey;
+};
+
+export const getDecryptedAttachment = async (
+    attachment: Attachment,
+    message: MessageExtended,
+    api: Api
+): Promise<AttachmentData> => {
+    const encryptedBinary = await getRequest(attachment, api);
+    try {
+        const sessionKey = await getSessionKey(attachment, message);
+        return await decrypt(encryptedBinary, sessionKey);
+    } catch (error) {
+        const blob = concatArrays([
+            binaryStringToArray(decodeBase64(attachment.KeyPackets)),
+            new Uint8Array(encryptedBinary)
+        ]);
+        // Fallback download raw attachment
+        throw { data: attachment, binary: blob, error };
+    }
+};
+
+export const getAndVerify = async (
```

Comment:
Since there are five arguments here, maybe it would be a bit more clear to use an object here. Like that it would be more clear what `false` refers to in the `attachmentDownloader`. But I ignore if there are some other reasons not to do that

Reply:
That could be a really vast debat. I asked myself the same question when designing the API of several helper of this project.

Here is how I see things:
- Using object as argument
  - `+` name parameters in pure JS
  - `+` can pass parameters in any order
  - `-` poor typing notations in TS for custom objects

```typescript
// object as arguments
const myFunc = ({arg1, arg2}: {arg1: string; arg2: number}) => 'result';
```

- Using several normal argument
  - `+` arguments are named with TS
  - `+` typing is a lot clearer to write
  - `+` (geeky one) easier to consider currying
  - `~` arguments are named in JS with the support of the TS server
  - `-` arguments have a fix order

```typescript
// standard arguments
const myFunc = (arg1: string, arg2: number) => 'result';
```

As you can see, I chose the second one but I clearly admit there is a debate.

There is a point I want to mention too. As I try to get out of Angular DI logic, I try to implement the most of the logic as pure functions helpers avoiding to use masked dependencies.
This, I hope, will tend to get things clearer and more testable. Yet it clearly increases the number or arguments of most functions.

Reply:
Normally we've been following the approach the Edu suggests, if there's more than 2 or 3 arguments the function can become quite convoluted. In these cases it's better to pass the options as an object, especially if some of the arguments are optional. Perhaps for the ts drawback we can declare it more as we do for the react components:

```ts
interface Props {
    arg1: string;
    arg2: number;
}

const myFunc = ({ arg1, arg2 }: Props) => {
```

Although I agree it doesn't improve it so much.

Reply:
I would use an interface too for objects. Yet for a React component alone in its file, it's easy but in a helper file with a dozen of functions, these arguments interface becomes a lot less nice to deal with.

Reply:
I think my comment had a more naive intention :sweat_smile: . It's simply that I found it ugly a bit ugly to have `false` in
```
const { data } = await getAndVerify(attachment, message, false, cache, api);
````
in the `attachmentDownloader` file, but an alternative solution would be simply
```
const reverify = false;
const { data } = await getAndVerify(attachment, message, false, cache, api);
```

Reply:
Ahah yes, it went a lot further ^^

But I'm happy to receive your opinions for my future APIs.

I'll use your proposition right now and don't change more because I suppose this will change in the future anyway.